File: ros2_ws/src/ugv_bringup_py/ugv_bringup_py/sensordebug_node.py
# This Node calculates covariances for Imu and MagneticField messages.
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Imu, MagneticField
from ament_index_python.packages import get_package_share_directory
import numpy as np
import json

class SensordebugNode(Node):
    def __init__(self):
        super().__init__('sensordebug_node')
        # declare parameters
        self.sub_name_imu = self.declare_parameter('sub_name_imu', 'imu/data_raw').value
        self.sub_name_mag = self.declare_parameter('sub_name_mag', 'imu/mag').value
        self.data_store_name = self.declare_parameter('data_store_name', 'covariances').value

        self.imu_acce_data_list = []
        self.imu_gyro_data_list = []
        self.mag_data_list = []
        self.count_imu = 0
        self.count_mag = 0
        self.sample_num = 1000
        self.data_save_path = "/home/nvidia/UGV_Rover/ros2_ws/src/ugv_bringup_py/config/"
        # ROS2 components
        # subs
        self.sub_imu = self.create_subscription(
            Imu,
            self.sub_name_imu,
            self.imu_callback,
            10)
        self.sub_mag = self.create_subscription(
            MagneticField,
            self.sub_name_mag,
            self.mag_callback,
            10)
        # timer
        self.timer = self.create_timer(1.0, self.timer_callback)
        # end of initialization
        self.get_logger().info("Sensordebug Node has been started.")

    def imu_callback(self, msg: Imu):
        if self.count_imu < self.sample_num:
            imu_data = [
                msg.linear_acceleration.x,
                msg.linear_acceleration.y,
                msg.linear_acceleration.z,
                msg.angular_velocity.x,
                msg.angular_velocity.y,
                msg.angular_velocity.z
            ]
            self.imu_acce_data_list.append(imu_data[:3])
            self.imu_gyro_data_list.append(imu_data[3:])
            self.count_imu += 1

    def mag_callback(self, msg: MagneticField):
        if self.count_mag < self.sample_num:
            mag_data = [
                msg.magnetic_field.x,
                msg.magnetic_field.y,
                msg.magnetic_field.z
            ]
            self.mag_data_list.append(mag_data)
            self.count_mag += 1

    def calc_result(self):
        imu_acce_array = np.array(self.imu_acce_data_list)
        imu_gyro_array = np.array(self.imu_gyro_data_list)
        mag_array = np.array(self.mag_data_list)

        imu_acce_cov = np.cov(imu_acce_array, rowvar=False)
        imu_gyro_cov = np.cov(imu_gyro_array, rowvar=False)
        mag_cov = np.cov(mag_array, rowvar=False)

        result = {
            'imu_acce_covariance': imu_acce_cov.flatten().tolist(),
            'imu_gyro_covariance': imu_gyro_cov.flatten().tolist(),
            'magnetic_field_covariance': mag_cov.flatten().tolist()
        }

        save_path = self.data_save_path + self.data_store_name + '.json'
        with open(save_path, 'w') as f:
            json.dump(result, f)

        self.get_logger().info(f"Covariance data saved to {save_path}")

    def timer_callback(self):
        if self.count_imu >= self.sample_num and self.count_mag >= self.sample_num:
            self.calc_result()
            self.timer.cancel()  # stop the timer after calculation
            self.destroy_node()
            rclpy.shutdown()
    
def main(args=None):
    rclpy.init(args=args)
    sensordebug_node = SensordebugNode()
    rclpy.spin(sensordebug_node)

File: ros2_ws/src/ugv_bringup_cpp/src/sensordebug_node.cpp
#include <array>
#include <fstream>
#include <memory>
#include <string>
#include <vector>

#include <nlohmann/json.hpp>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/magnetic_field.hpp>

namespace ugv_bringup_cpp {

class SensordebugNode : public rclcpp::Node {
public:
  SensordebugNode() : Node("sensordebug_node") {
    sub_name_imu_ = this->declare_parameter<std::string>("sub_name_imu", "imu/data_raw");
    sub_name_mag_ = this->declare_parameter<std::string>("sub_name_mag", "imu/mag");
    data_store_name_ = this->declare_parameter<std::string>("data_store_name", "covariances");
    data_save_path_ = this->declare_parameter<std::string>(
        "data_save_path", "/home/nvidia/UGV_Rover/ros2_ws/src/ugv_bringup_py/config/");
    sample_num_ = this->declare_parameter<int>("sample_num", 1000);

    sub_imu_ = this->create_subscription<sensor_msgs::msg::Imu>(
        sub_name_imu_, 10, std::bind(&SensordebugNode::ImuCallback, this, std::placeholders::_1));
    sub_mag_ = this->create_subscription<sensor_msgs::msg::MagneticField>(
        sub_name_mag_, 10, std::bind(&SensordebugNode::MagCallback, this, std::placeholders::_1));

    timer_ = this->create_wall_timer(std::chrono::seconds(1),
                                     std::bind(&SensordebugNode::TimerCallback, this));
    RCLCPP_INFO(this->get_logger(), "Sensordebug Node has been started.");
  }

private:
  void ImuCallback(const sensor_msgs::msg::Imu::SharedPtr msg) {
    if (count_imu_ >= sample_num_) {
      return;
    }
    imu_acce_data_.push_back({msg->linear_acceleration.x, msg->linear_acceleration.y,
                              msg->linear_acceleration.z});
    imu_gyro_data_.push_back({msg->angular_velocity.x, msg->angular_velocity.y, msg->angular_velocity.z});
    count_imu_++;
  }

  void MagCallback(const sensor_msgs::msg::MagneticField::SharedPtr msg) {
    if (count_mag_ >= sample_num_) {
      return;
    }
    mag_data_.push_back({msg->magnetic_field.x, msg->magnetic_field.y, msg->magnetic_field.z});
    count_mag_++;
  }

  void TimerCallback() {
    if (count_imu_ >= sample_num_ && count_mag_ >= sample_num_) {
      CalcResult();
      timer_->cancel();
      rclcpp::shutdown();
    }
  }

  static std::vector<double> Covariance(const std::vector<std::array<double, 3>> &data) {
    std::vector<double> cov(9, 0.0);
    if (data.empty()) {
      return cov;
    }
    double mean[3] = {0.0, 0.0, 0.0};
    for (const auto &row : data) {
      mean[0] += row[0];
      mean[1] += row[1];
      mean[2] += row[2];
    }
    mean[0] /= data.size();
    mean[1] /= data.size();
    mean[2] /= data.size();
    for (const auto &row : data) {
      double dx = row[0] - mean[0];
      double dy = row[1] - mean[1];
      double dz = row[2] - mean[2];
      cov[0] += dx * dx;
      cov[1] += dx * dy;
      cov[2] += dx * dz;
      cov[3] += dy * dx;
      cov[4] += dy * dy;
      cov[5] += dy * dz;
      cov[6] += dz * dx;
      cov[7] += dz * dy;
      cov[8] += dz * dz;
    }
    double denom = data.size() > 1 ? static_cast<double>(data.size() - 1) : 1.0;
    for (auto &val : cov) {
      val /= denom;
    }
    return cov;
  }

  void CalcResult() {
    auto imu_acce_cov = Covariance(imu_acce_data_);
    auto imu_gyro_cov = Covariance(imu_gyro_data_);
    auto mag_cov = Covariance(mag_data_);

    nlohmann::json result;
    result["imu_acce_covariance"] = imu_acce_cov;
    result["imu_gyro_covariance"] = imu_gyro_cov;
    result["magnetic_field_covariance"] = mag_cov;

    std::string save_path = data_save_path_ + data_store_name_ + ".json";
    std::ofstream out(save_path);
    if (!out.is_open()) {
      RCLCPP_ERROR(this->get_logger(), "Failed to open output file: %s", save_path.c_str());
      return;
    }
    out << result.dump();
    RCLCPP_INFO(this->get_logger(), "Covariance data saved to %s", save_path.c_str());
  }

  std::string sub_name_imu_;
  std::string sub_name_mag_;
  std::string data_store_name_;
  std::string data_save_path_;
  int sample_num_{1000};

  int count_imu_{0};
  int count_mag_{0};
  std::vector<std::array<double, 3>> imu_acce_data_;
  std::vector<std::array<double, 3>> imu_gyro_data_;
  std::vector<std::array<double, 3>> mag_data_;

  rclcpp::Subscription<sensor_msgs::msg::Imu>::SharedPtr sub_imu_;
  rclcpp::Subscription<sensor_msgs::msg::MagneticField>::SharedPtr sub_mag_;
  rclcpp::TimerBase::SharedPtr timer_;
};

}  // namespace ugv_bringup_cpp

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<ugv_bringup_cpp::SensordebugNode>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
